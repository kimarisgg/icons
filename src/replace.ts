import icons from "./icons"

function replace() {
    if (typeof document === "undefined") {
        throw new Error("`replace()` only works in a browser environment.")
    }

    document.querySelectorAll("[data-kimaris]").forEach((element) => replaceElement(element))
}

function replaceElement(element: Element) {
    const elementAttrs = element.attributes
    const name = elementAttrs.getNamedItem("data-kimaris")!!
    elementAttrs.removeNamedItem("data-kimaris")
    if (icons.has(name.value)) {
        const svg = icons.get(name.value)!!.toSvg(elementAttrs)
        element.replaceWith(svg)
    }
}

export default replace
