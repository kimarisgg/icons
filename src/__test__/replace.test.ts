import replace from "../replace"

jest.mock('../generated/icons.json', () => ([
    {
        name: "icon1",
        type: "common",
        value: '<line x1="23" y1="1" x2="1" y2="23" /><line x1="1" y1="1" x2="23" y2="23" />',
    },
    {
        name: "icon2",
        type: "common",
        value: '<circle cx="12" cy="12" r="11" />',
    }
]));

test("replaces [data-kimaris] elements with SVG markup", () => {
    document.body.innerHTML = '<i data-kimaris="icon1"></i><span data-kimaris="icon2"></i>'
    expect(document.body.innerHTML).toMatchSnapshot()
    replace()
    expect(document.body.innerHTML).toMatchSnapshot()
})

test("copies placeholder element attributes to <svg> tag", () => {
    document.body.innerHTML = '<i data-kimaris="icon1" id="test" class="foo bar" stroke-width="1"></i>'
    expect(document.body.innerHTML).toMatchSnapshot()
    replace()
    expect(document.body.innerHTML).toMatchSnapshot()
})
