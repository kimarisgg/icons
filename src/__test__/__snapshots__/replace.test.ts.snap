// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copies placeholder element attributes to <svg> tag 1`] = `"<i data-kimaris="icon1" id="test" class="foo bar" stroke-width="1"></i>"`;

exports[`copies placeholder element attributes to <svg> tag 2`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewbox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1" stroke-linecap="round" stroke-linejoin="round" id="test" class="foo bar"><line x1="23" y1="1" x2="1" y2="23"><line x1="1" y1="1" x2="23" y2="23"></line></line></svg>"`;

exports[`replaces [data-kimaris] elements with SVG markup 1`] = `"<i data-kimaris="icon1"></i><span data-kimaris="icon2"></span>"`;

exports[`replaces [data-kimaris] elements with SVG markup 2`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewbox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><line x1="23" y1="1" x2="1" y2="23"><line x1="1" y1="1" x2="23" y2="23"></line></line></svg><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewbox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><circle cx="12" cy="12" r="11"></circle></svg>"`;
